### ISSUE TITLE
TypeError: Cannot read property 'then' of undefined

### ISSUE THREAD

**Issue by dabbers** [phillipj/node-plex-api#72]
The plex api crashes when authenticating (I'm assuming), and I see this stack trace:

> < src\node_modules\plex-api\lib\api.js:174
> <                 .then(function () {
> <                 ^
> < TypeError: Cannot read property 'then' of undefined
> <     at Request.onResponse [as _callback](srcnode_modulesplex-apilibapi.js:174:17)
> <     at Request.self.callback (src\node_modules\plex-api\node_modules\request\request.js:200:22)
> <     at emitTwo (events.js:106:13)
> <     at Request.emit (events.js:191:7)
> <     at Request.<anonymous> (src\node_modules\plex-api\node_modules\request\request.js:1046:10)
> <     at emitOne (events.js:101:20)
> <     at Request.emit (events.js:188:7)
> <     at IncomingMessage.<anonymous> (src\node_modules\plex-api\node_modules\request\request.js:973:12)
> <     at emitNone (events.js:91:20)
> <     at IncomingMessage.emit (events.js:185:7)

Code snippet: 

```
      var client = new PlexAPI({
        hostname: Core.config.Modules.PlexAnnounce.hostname,
        port:     Core.config.Modules.PlexAnnounce.port,
        username: Core.config.Modules.PlexAnnounce.username,
        password: Core.config.Modules.PlexAnnounce.password,
        options: {
            identifier: "6BC406D9-AB71-48B4-B5C1-90C2458E5396",
            product: "DeBot.js Module"
        }
    });

    client.query("/library/recentlyAdded").then(function (result) {
                 // code stuff here... (never gets executed)
    }, function (err) {
        console.log("[PlexAnnounce] failed to connect to plex server....", err); // Also doesn't get executed
    });
```

This code worked at one point, I moved the server, then it started exhibiting this behavior. It was working last week, then it stopped working again today. Not sure what the root cause is, but I'm not familiar enough with the API gateway/API auth to understand the issue.

**Comment by phillipj** [comment]
Oh my, this doesn't look good. Thanks for reporting, I'll have a look and
get back to you

**Comment by phillipj** [comment]
@dabbers at first glance, I cant see the reason how this could happen tbh. Some extra information would be most valuable.

When you say you moved the server, which version of Node.js and OS were you running on those servers?

How did you move your application? E.g. copy the directory whole application directory including the `node_modules` directory, and run it directly? Or did you run `$ npm install` on the new server?

**Comment by dabbers** [comment]
When I say I moved the server, I was referring to the plex server. It used to be locally hosted, but now it's externally hosted. Port is forwarded, I can access plex server from externally. The application itself has stayed the same, unchanged (minus the host and port). 

I suppose the node.js version might have changed recently. Though the module was broken before the version update too. 

C:\Users\me>node -v
v6.6.0

I think before I was using 4.4.0 or 4.6.0, but it was still an issue.

If there's anything you need let me know. I can run a logged version of your library if you think that'll be helpful. 

I've decided to run wireshark to see the requests, and it looks like I'm receiving an 401 response, even though I know for a fact the credentials are correct. 

![2016-10-12_1422](https://cloud.githubusercontent.com/assets/480822/19328184/5993ef5a-9087-11e6-963c-24a536156666.png)
![2016-10-12_1423](https://cloud.githubusercontent.com/assets/480822/19328247/a3c0c22e-9087-11e6-8c0e-f6182ce939f4.png)

**Comment by dabbers** [comment]
Further investigation, it might be related to an issue on my end. We seem to have 2 plex servers hosted there, the one I log into is the one I want because it's been port forwarded, but it appears as unavailable for a few seconds and defaults me to the other server, which has been shared with me, so I'm not sure why the api returns 401. I don't think this issue I'm seeing is with the API, though it would be nice to get some extra error checking regarding this error path.

**Comment by phillipj** [comment]
> I don't think this issue I'm seeing is with the API, though it would be nice to get some extra error checking regarding this error path.

Completely agree! No matter how the Plex Server responds or fails, the plex-api library shouldn't crash like what you've reported, and of course let you `.catch()` in any error situation. I still haven't been able to reproduce this behaviour tho, which is really annoying.

**Comment by mistic100** [comment]
I have the same issue when I query the server too often.

My script is very simple, it displays the title of the current played file : 

```js
var PlexAPI = require("plex-api");

var client = new PlexAPI({
  hostname: "127.0.0.1",
  username: '...',
  password: '...'
});

client.query({
  uri: '/status/sessions', 
  extraHeaders: {
    'Accept': 'application/xml'
  }
}).then(function(result) {
  console.log(result.MediaContainer.Track[0].attributes.grandparentTitle);
  console.log(result.MediaContainer.Track[0].attributes.parentTitle);
  console.log(result.MediaContainer.Track[0].attributes.title);
});
```

If I run this script more than once every 5 seconds (kinda) I get :

```
D:\rainmeter-plex\node_modules\plex-api\lib\api.js:181
                .then(function () {
                ^

TypeError: Cannot read property 'then' of undefined
    at Request.onResponse [as _callback] (D:\rainmeter-plex\node_modules\plex-api\lib\api.js:181:17)
    at Request.self.callback (D:\rainmeter-plex\node_modules\request\request.js:200:22)
    at emitTwo (events.js:106:13)
    at Request.emit (events.js:191:7)
    at Request.<anonymous> (D:\rainmeter-plex\node_modules\request\request.js:1067:10)
    at emitOne (events.js:101:20)
    at Request.emit (events.js:188:7)
    at IncomingMessage.<anonymous> (D:\rainmeter-plex\node_modules\request\request.js:988:12)
    at emitNone (events.js:91:20)
    at IncomingMessage.emit (events.js:185:7)
    at endReadableNT (_stream_readable.js:974:12)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9)
```

I guess I should not re-authenticate every time and in the final program I will not, but it's very odd to crash like this.

**Comment by dabbers** [comment]
I've created a simple repro for this. 

I start up a node.js server like so:

```
var http = require('http');
var fs = require('fs');

http.createServer(function (req, res) {
    console.log(req.url);
  res.writeHead(401, {'Content-Type': 'text/plain'});
  res.end("401 unauthorized");
}).listen(32400);
```


Then I connect to this nodd.js web server. 

```
var PlexAPI = require("plex-api");

var client = new PlexAPI({
	hostname: "192.168.0.102",
	port: 32400,
	username: "blah",
	password: "blah"
});
client.query("/library/recentlyAdded").then(function (res) {
console.log("res", res);
}, function (er) {
console.log("error", er);
});
```


Hopefully this helps diagnose how you can be more resilient to strange errors from Plex.

**Comment by phillipj** [comment]
Much appreciated! 👍

**Comment by hipstersmoothie** [comment]
experiencing this issue as well.

**Comment by dabbers** [comment]
I found that if the account has multiple servers shared to it, it will go to another server from time to time. I'm guessing the api is authenticating with one server, then switching to another server somehow. I don't pretend to understand how the api works with multiple servers on the account, but when I removed the other servers from the account, the api worked again.  

Just sharing my workaround.

**Comment by phillipj** [comment]
The previous example in https://github.com/phillipj/node-plex-api/issues/72#issuecomment-263157386 made it really simple to reproduce. I just opened https://github.com/phillipj/node-plex-api/issues/76 which seemed to fix the issue.

I'll keep that PR open for a day or two in case anyone has any thoughts or improvements, and merge / publish a new plex-api package version after that.

**Comment by phillipj** [comment]
@dabbers I just published v5 of this package, which should fix the `undefined` issue you've experienced.

Beware of the breaking change mentioned in #76 & #77, feel free to ask if there's any questions.

**Comment by dabbers** [comment]
Thanks! I resolved the issue for now by removing the other servers shared on my account. I'll give it a test run later on though. 

Thank you!

**Comment by dabbers** [comment]
Just to follow through with this issue, I'm closer to understanding the issue with the 401 error. It's related to the auth token. I noticed that it was sending the same token every time, but other auth tokens would work. So to fix, I changed the unique identifier (though I suspect removing the client from plex would work too). I'm not sure why this is happening where the API token becomes invalid, even though the plex.tv api always returns the key for the unique ID and username/pw combo. I suspect the server might be doing some rate throttling but I don't know if that's the case. 

Anyway, in case anyone happens upon this issue too, maybe this will help them. I've added a small math.random bit of code to my unique identifier every so often to help keep the unique id from getting blocked. I just need to make sure to go through old clients in the web api to keep it tidy.
